File: docs/chapters/chapter-5/assets/case-THU-DA-L02-C01-A02/python/quality.py
"""显式教学规则：保留原始字符，检查后生成派生表，不改原始文件。"""
from pathlib import Path
import argparse
import csv
import re
import pandas as pd
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt

BASE = Path(__file__).resolve().parents[1]
COLUMNS = ["record_id", "participant_id", "round", "estimate", "unit"]

def clean(source):
    with Path(source).open(encoding="utf-8-sig", newline="") as handle:
        reader = csv.DictReader(handle)
        if reader.fieldnames != COLUMNS:
            raise ValueError("Expected exact column names and order")
        rows = list(reader)
    if any(None in row or any(v is None for v in row.values()) for row in rows):
        raise ValueError("Malformed row")
    if any(not row["record_id"] or not row["participant_id"] or row["round"] not in ("1", "2") for row in rows):
        raise ValueError("Missing ID or invalid round; stop for review")
    seen, unique, issues = {}, [], []
    for number, row in enumerate(rows, 1):
        rid = row["record_id"]
        if rid in seen:
            if row != seen[rid]:
                raise ValueError("Conflicting record_id: " + rid)
            issues.append([number, rid, "duplicate_export", "drop_duplicate_copy"])
        else:
            seen[rid] = row
            unique.append((number, row))
    output, quarantine = [], []
    for number, row in unique:
        raw, unit = row["estimate"], row["unit"]
        issue = ""
        if raw == "":
            issue = "missing_estimate"
        elif not re.fullmatch(r"[+-]?(?:[0-9]+(?:\.[0-9]*)?|\.[0-9]+)", raw):
            issue = "non_numeric"
        elif unit not in ("粒", "千粒"):
            issue = "unknown_unit"
        else:
            value = float(raw) * (1000 if unit == "千粒" else 1)
            if value < 0:
                issue = "negative_estimate"
        if issue:
            issues.append([number, row["record_id"], issue, "quarantine"])
            quarantine.append({**row, "reason": issue})
            continue
        high = value > 5000
        if high:
            issues.append([number, row["record_id"], "high_review", "retain_flag"])
        output.append([row["record_id"], row["participant_id"], int(row["round"]), value, "TRUE" if high else "FALSE"])
    result = pd.DataFrame(output, columns=["record_id","participant_id","round","estimate_grains","review_high"])
    flow = pd.DataFrame([
        ["raw",len(rows),len({r["participant_id"] for r in rows})],
        ["deduplicated",len(unique),len({r["participant_id"] for _,r in unique})],
        ["eligible",len(result),result.participant_id.nunique()]
    ], columns=["stage","records","participants"])
    return result, pd.DataFrame(issues,columns=["source_row","record_id","issue","action"]), pd.DataFrame(quarantine,columns=COLUMNS+["reason"]), flow

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--input", type=Path, default=BASE/"data/guesses_raw.csv")
    parser.add_argument("--out", type=Path, default=BASE/"results/python")
    args = parser.parse_args()
    result, issues, quarantine, flow = clean(args.input)
    args.out.mkdir(parents=True, exist_ok=True)
    for name, frame in [("clean",result),("issues",issues),("quarantine",quarantine),("flow",flow)]:
        frame.to_csv(args.out/(name+".csv"),index=False,encoding="utf-8")
    fig, ax = plt.subplots(figsize=(6.4,3.6), layout="constrained")
    bars = ax.bar(flow.stage, flow.records, color=["#8a949d","#5c8a99","#276675"],width=0.55)
    ax.bar_label(bars)
    ax.set(ylabel="Number of records", ylim=(0,16), title="Teaching data: processing stages")
    ax.spines[["top","right"]].set_visible(False)
    for suffix in ("png","svg"):
        fig.savefig(args.out/("quality-flow."+suffix), dpi=160)
        if args.out.resolve() == (BASE/"results/python").resolve():
            (BASE/"figures").mkdir(exist_ok=True)
            fig.savefig(BASE/"figures"/("quality-flow."+suffix), dpi=160)
    plt.close(fig)
    print(flow.to_string(index=False))

if __name__ == "__main__":
    main()

File: docs/chapters/chapter-5/assets/case-THU-DA-L02-C01-A02/r/quality.R
# 基础R；输入、规则、保留顺序与Python一致。
options(warn=2)
if (.Platform$OS.type=="windows" && !l10n_info()[["UTF-8"]]) {
  invisible(Sys.setlocale("LC_CTYPE","English_United States.utf8"))
}
script <- sub("^--file=", "", grep("^--file=", commandArgs(), value=TRUE)[1])
base <- dirname(dirname(normalizePath(script, mustWork=TRUE)))
args <- commandArgs(trailingOnly=TRUE)
source <- if (length(args)>=1) args[1] else file.path(base,"data/guesses_raw.csv")
out <- if (length(args)>=2) args[2] else file.path(base,"results/r")
columns <- c("record_id","participant_id","round","estimate","unit")
frame <- read.csv(source, colClasses="character", na.strings=NULL, fileEncoding="UTF-8", check.names=FALSE, blank.lines.skip=FALSE)
if (!identical(names(frame),columns)) stop("Expected exact column names and order")
if (anyNA(frame) || any(frame$record_id=="") || any(frame$participant_id=="") || any(!frame$round %in% c("1","2"))) stop("Missing ID or invalid round; stop for review")
for (rid in unique(frame$record_id)) {
  block <- frame[frame$record_id==rid,,drop=FALSE]
  if (nrow(unique(block))>1) stop(paste("Conflicting record_id:",rid))
}
dup <- duplicated(frame)
issues <- data.frame(source_row=which(dup),record_id=frame$record_id[dup],
                     issue=rep("duplicate_export",sum(dup)),action=rep("drop_duplicate_copy",sum(dup)))
source_rows <- which(!dup)
unique_frame <- frame[!dup,,drop=FALSE]
result <- data.frame(record_id=character(),participant_id=character(),round=integer(),estimate_grains=numeric(),review_high=character())
quarantine <- unique_frame[FALSE,,drop=FALSE]
quarantine$reason <- character()
for (j in seq_len(nrow(unique_frame))) {
  row <- unique_frame[j,,drop=FALSE]
  issue <- ""
  if (row$estimate=="") {
    issue <- "missing_estimate"
  } else if (!grepl("^[+-]?([0-9]+(\\.[0-9]*)?|\\.[0-9]+)$",row$estimate)) {
    issue <- "non_numeric"
  } else if (!row$unit %in% c("粒","千粒")) {
    issue <- "unknown_unit"
  } else {
    value <- as.numeric(row$estimate) * if (row$unit=="千粒") 1000 else 1
    if (value<0) issue <- "negative_estimate"
  }
  if (nzchar(issue)) {
    issues <- rbind(issues,data.frame(source_row=source_rows[j],record_id=row$record_id,issue=issue,action="quarantine"))
    row$reason <- issue
    quarantine <- rbind(quarantine,row)
    next
  }
  high <- value>5000
  if (high) issues <- rbind(issues,data.frame(source_row=source_rows[j],record_id=row$record_id,issue="high_review",action="retain_flag"))
  result <- rbind(result,data.frame(record_id=row$record_id,participant_id=row$participant_id,round=as.integer(row$round),estimate_grains=value,review_high=if (high) "TRUE" else "FALSE"))
}
flow <- data.frame(stage=c("raw","deduplicated","eligible"),
                  records=c(nrow(frame),nrow(unique_frame),nrow(result)),
                  participants=c(length(unique(frame$participant_id)),length(unique(unique_frame$participant_id)),length(unique(result$participant_id))))
dir.create(out,recursive=TRUE,showWarnings=FALSE)
for (name in c("clean","issues","quarantine","flow")) {
  object <- switch(name,clean=result,issues=issues,quarantine=quarantine,flow=flow)
  write.csv(object,file.path(out,paste0(name,".csv")),row.names=FALSE,na="",fileEncoding="UTF-8")
}
draw <- function() {
  mids <- barplot(flow$records,names.arg=flow$stage,col=c("#8a949d","#5c8a99","#276675"),
                  ylim=c(0,16),ylab="Number of records",main="Teaching data: processing stages",border=NA)
  text(mids,flow$records+0.6,labels=flow$records)
}
png(file.path(out,"quality-flow.png"),width=1024,height=576,res=160)
draw()
dev.off()
svg(file.path(out,"quality-flow.svg"),width=6.4,height=3.6)
draw()
dev.off()
print(flow)
